test/common/buffer/buffer_test.cc

```diff
@@ -14,6 +14,252 @@ namespace Envoy {
 namespace Buffer {
 namespace {
 
+class DummySlice : public Slice {
+public:
+  DummySlice(const std::string& data, const std::function<void()>& deletion_callback)
+      : Slice(0, data.size(), data.size()), deletion_callback_(deletion_callback) {
+    base_ = reinterpret_cast<uint8_t*>(const_cast<char*>(data.c_str()));
+  }
+  ~DummySlice() override {
+    if (deletion_callback_ != nullptr) {
+      deletion_callback_();
+    }
+  }
+
+private:
+  const std::function<void()> deletion_callback_;
+};
+
+bool sliceMatches(const SlicePtr& slice, const std::string& expected) {
+  return slice != nullptr && slice->dataSize() == expected.size() &&
+         memcmp(slice->data(), expected.c_str(), expected.size()) == 0;
+}
+
+TEST(OwnedSliceTest, Create) {
+  static std::vector<uint64_t> Sizes = {0, 1, 64, 4096 - sizeof(OwnedSlice), 65535};
+  for (const auto size : Sizes) {
+    auto slice = OwnedSlice::create(size);
+    EXPECT_NE(nullptr, slice->data());
+    EXPECT_EQ(0, slice->dataSize());
+    EXPECT_LE(size, slice->reservableSize());
+  }
+}
+
+TEST(OwnedSliceTest, ReserveCommit) {
+  auto slice = OwnedSlice::create(100);
+  const uint64_t initial_capacity = slice->reservableSize();
+  EXPECT_LE(100, initial_capacity);
+
+  {
+    // Verify that a zero-byte reservation is rejected.
+    Slice::Reservation reservation = slice->reserve(0);
+    ASSERT_EQ(nullptr, reservation.mem_);
+    ASSERT_EQ(0, reservation.len_);
+    ASSERT_EQ(initial_capacity, slice->reservableSize());
+  }
+
+  {
+    // Create a reservation smaller than the reservable size.
+    // It should reserve the exact number of bytes requested.
+    Slice::Reservation reservation = slice->reserve(10);
+    ASSERT_EQ(slice->data(), reservation.mem_);
+    ASSERT_EQ(10, reservation.len_);
+
+    // Request a second reservation while the first reservation remains uncommitted.
+    // This should fail.
+    ASSERT(slice->reservableSize() == 0);
+    Slice::Reservation reservation2 = slice->reserve(1);
+    ASSERT_EQ(nullptr, reservation2.mem_);
+    ASSERT_EQ(0, reservation2.len_);
+
+    // Commit the entire reserved size.
+    bool committed = slice->commit(reservation);
+    ASSERT(committed);
+    ASSERT_EQ(10, slice->dataSize());
+    ASSERT_EQ(initial_capacity - 10, slice->reservableSize());
+
+    // Verify that a reservation can only be committed once.
+    ASSERT_FALSE(slice->commit(reservation));
```

Comment:
it looks like these stanzas have a lot of common code...could you consider making a test-class and putting a helper-method in there so there's fewer/smaller repeated blocks of code?